### DISCUSSION TITLE
Explanation of PreInitialization, Language Instance, Contexts in the TruffleRuby language implementation

### DISCUSSION THREAD

**Discussion by petenorth** [oracle/truffleruby#2323] [General] [answered]
Following on from Christian Humer publishing

https://github.com/oracle/graal/blob/master/truffle/docs/AOTOverview.md

It would be good if there was a high level explanation of when new `RubyContext`s and `RubyLanguage` instances are created.

Should this describe any language instances or contexts associated with llvm and when they are created?

A real example program would help i.e. rails / rspec / etc

Maybe go on to describe the current situation and consequence for deoptimsations.

And then what the ultimate goal is (I guess referencing #1835 )

I for one am curious, maybe there are others also?

**Comment by eregon** [comment, marked as answer]
I'm sorry for the late answer, I missed this, the Slack GitHub plugin finally added support for Discussions.

> It would be good if there was a high level explanation of when new `RubyContext`s and `RubyLanguage` instances are created.

A RubyContext is an execution context / isolate, e.g. each execution context has its own independent state, for instance its own set of global variables, constants, classes, etc.
The only objects which are/cam be shared between contexts are deeply immutable objects like Symbols, frozen string literals, nil, true, false, Regexps, Integers, Float, etc.

A RubyLanguage is what holds onto these are deeply immutable objects and on the AST, so on the representation of the code. So multiple RubyContexts sharing the same RubyLanguage e.g. only need to compile a method once per RubyLanguage and not once per RubyContext.

> Should this describe any language instances or contexts associated with llvm and when they are created?

Context and language instances always match, so 1 LLVMContext per RubyContext, and the same for languages.

> A real example program would help i.e. rails / rspec / etc

Whenever you run using `ruby` or `truffleruby` (or any gem executable which uses those) it creates a single RubyContext and a single RubyLanguage.
It's possible to create inner contexts with the Polyglot::InnerContext API: https://github.com/oracle/truffleruby/blob/7cc607338c2d83cb834f3f5afe0016af6e523f08/src/main/ruby/truffleruby/core/truffle/polyglot.rb#L16
Or with the Java [`Context`](https://www.graalvm.org/sdk/javadoc/org/graalvm/polyglot/Context.html) API, or with the C/C++ APIs which interacts with the Java API.

> Maybe go on to describe the current situation and consequence for deoptimsations.

Sharing the RubyLanguage for multiple contexts is currently not supported by default for TruffleRuby because it's not fast yet (e.g., uncached method and constant lookups). #1835 tracks that.
So what happens currently is there is one RubyLanguage per RubyContext, but so no AST sharing between RubyContexts.

> And then what the ultimate goal is (I guess referencing #1835)

To share code between multiple contexts but also with Auxiliary Engine Caching (EE) to be able to persist the JITed code on disk and reuse for the next process, even when using a single context as with the `ruby`/`truffleruby` launchers.
